File: services/api/internal/repository/finding_order_test.go
package repository

import (
	"encoding/json"
	"sort"
	"strings"
	"testing"

	"github.com/aegis-platform/api/internal/models"
)

// B1 parity (P4b): the Go SQL ordering and the web comparator
// (web/lib/findingOrder.ts) must agree, or page 2 shows a finding that belonged on
// page 1. This asserts the SQL-key order on the CANONICAL fixture matches the same
// order the TS test asserts (findingOrder.test.ts). Both reference one spec, so a
// drift on either side fails its own test.

func fp(v float64) *float64 { return &v }
func sp(v string) *string   { return &v }

func orderFinding(id, severity string, kev, reachable, isNew bool, fpProb float64) models.Finding {
	m := map[string]any{}
	if kev {
		m["kev"] = true
	}
	if reachable {
		m["reachable"] = true
	}
	raw, _ := json.Marshal(m)
	return models.Finding{
		RuleID: id, Severity: severity, IsNew: isNew,
		FalsePositiveProbability: fp(fpProb), Fingerprint: sp("f-" + id), Metadata: models.JSONB(raw),
	}
}

// sqlKey mirrors, field for field, the ORDER BY expressions in orderByFindings:
// kevFirstSQL, severityRankSQL, reachableFirstSQL, newFirstSQL, fp, fingerprint. It
// reads metadata the way `metadata->>'kev' = 'true'` does — from the raw JSON.
func sqlKey(f models.Finding) []int {
	b := func(v bool) int {
		if v {
			return 0
		}
		return 1
	}
	sevRank := map[string]int{"critical": 0, "high": 1, "medium": 2, "low": 3}
	sr, ok := sevRank[f.Severity]
	if !ok {
		sr = 4
	}
	var meta map[string]any
	if len(f.Metadata) > 0 {
		_ = json.Unmarshal(f.Metadata, &meta)
	}
	kev := meta["kev"] == true
	reach := meta["reachable"] == true
	return []int{b(kev), sr, b(reach), b(f.IsNew)}
}

func TestFindingOrderMatchesCanonicalSpec(t *testing.T) {
	// Same fixture + expected order as web/lib/findingOrder.test.ts.
	fixture := []models.Finding{
		orderFinding("low", "low", false, false, false, 0),
		orderFinding("high-fp", "high", false, false, false, 0.9),
		orderFinding("kev-crit", "critical", true, false, false, 0),
		orderFinding("high", "high", false, false, false, 0),
		orderFinding("reach-crit", "critical", false, true, false, 0),
		orderFinding("high-new", "high", false, false, true, 0),
		orderFinding("crit", "critical", false, false, false, 0),
		orderFinding("reach-high", "high", false, true, false, 0),
	}
	want := []string{"kev-crit", "reach-crit", "crit", "reach-high", "high-new", "high", "high-fp", "low"}

	sort.SliceStable(fixture, func(i, j int) bool {
		ki, kj := sqlKey(fixture[i]), sqlKey(fixture[j])
		for x := range ki {
			if ki[x] != kj[x] {
				return ki[x] < kj[x]
			}
		}
		fi := *fixture[i].FalsePositiveProbability
		fj := *fixture[j].FalsePositiveProbability
		if fi != fj {
			return fi < fj
		}
		return *fixture[i].Fingerprint < *fixture[j].Fingerprint
	})

	got := make([]string, len(fixture))
	for i, f := range fixture {
		got[i] = f.RuleID
	}
	for i := range want {
		if got[i] != want[i] {
			t.Fatalf("order mismatch at %d: got %v, want %v", i, got, want)
		}
	}
}

// Structural guard: the single canonical ORDER BY keeps the four ranking keys in the
// exact tier sequence the comparator relies on. If someone reorders the SQL (e.g.
// drops reachable), this fails even before a data test would.
func TestOrderByRankingSequence(t *testing.T) {
	kev := strings.Index(orderByFindings, "metadata->>'kev'")
	sev := strings.Index(orderByFindings, "CASE severity")
	reach := strings.Index(orderByFindings, "metadata->>'reachable'")
	isNew := strings.Index(orderByFindings, "is_new")
	fpIdx := strings.Index(orderByFindings, "false_positive_probability")
	for _, p := range []int{kev, sev, reach, isNew, fpIdx} {
		if p < 0 {
			t.Fatalf("orderByFindings missing a ranking key: %q", orderByFindings)
		}
	}
	if !(kev < sev && sev < reach && reach < isNew && isNew < fpIdx) {
		t.Fatalf("ranking keys out of order in orderByFindings: kev=%d sev=%d reach=%d new=%d fp=%d", kev, sev, reach, isNew, fpIdx)
	}
}

File: web/lib/findingOrder.test.ts
// B1 parity (P4b): the client comparator must reproduce the exact order the API SQL
// returns, or page 2 shows a finding that belonged on page 1 — a silent bug. This
// asserts the client side against the CANONICAL fixture + order; the Go side
// (services/api/internal/repository/finding_order_test.go) asserts the SQL keys
// against the same fixture. Both reference the same spec, so a drift on either side
// fails its test. Run: node --test --experimental-strip-types lib/*.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { compareFindings, type OrderableFinding } from "./findingOrder.ts";

// Canonical fixture — shared with the Go parity test. Scrambled input; each id
// encodes its ordering-relevant fields.
const FIXTURE: (OrderableFinding & { id: string })[] = [
  { id: "low", severity: "low", fingerprint: "f-low" },
  { id: "high-fp", severity: "high", false_positive_probability: 0.9, fingerprint: "f-hf" },
  { id: "kev-crit", severity: "critical", metadata: { kev: true }, fingerprint: "f-kev" },
  { id: "high", severity: "high", fingerprint: "f-h" },
  { id: "reach-crit", severity: "critical", metadata: { reachable: true }, fingerprint: "f-rc" },
  { id: "high-new", severity: "high", is_new: true, fingerprint: "f-hn" },
  { id: "crit", severity: "critical", fingerprint: "f-c" },
  { id: "reach-high", severity: "high", metadata: { reachable: true }, fingerprint: "f-rh" },
];

// KEV → severity → reachable → new → FP-down → fingerprint.
const CANONICAL_ORDER = [
  "kev-crit",   // KEV outranks everything
  "reach-crit", // critical, reachable
  "crit",       // critical
  "reach-high", // high, reachable
  "high-new",   // high, new
  "high",       // high
  "high-fp",    // high, likely-FP pushed down
  "low",        // low last
];

test("client comparator matches the canonical (server) order", () => {
  const sorted = [...FIXTURE].sort(compareFindings).map((f) => f.id);
  assert.deepEqual(sorted, CANONICAL_ORDER);
});

test("ordering is deterministic (stable across a re-sort)", () => {
  const once = [...FIXTURE].sort(compareFindings).map((f) => f.id);
  const twice = [...FIXTURE].sort(compareFindings).sort(compareFindings).map((f) => f.id);
  assert.deepEqual(once, twice);
});
